File: intervalListIntersections.py
class Solution(object):
    def intervalIntersection(self, A, B):
        """
        :type A: List[List[int]]
        :type B: List[List[int]]
        :rtype: List[List[int]]
        """
        def isOverlap(A, B):
            return not (A[1] < B[0] or A[0] > B[1])
        def mostCommon(A, B):
            return [max(A[0],B[0]), min(A[1], B[1])]
        
        if not A or not B: return []
        la, lb = len(A), len(B)
        a, b = 0, 0
        ans = []
        while a < la and b < lb:
            if isOverlap(A[a], B[b]):
                ans.append(mostCommon(A[a], B[b]) )
            if A[a][1] > B[b][1]:
                b += 1
            else:
                a += 1
        return ans

File: intervalListIntersections.kt
class Solution {
    fun intervalIntersection(firstList: Array<IntArray>, secondList: Array<IntArray>): Array<IntArray> {
        var firstIdx = 0
        var secIdx = 0
        var result = mutableListOf<IntArray>()
        while (firstIdx < firstList.size && secIdx < secondList.size) {
            val overLap = getOverLap(firstList[firstIdx], secondList[secIdx])
            if (overLap != null) result += overLap!!
            //move up first or second index
            if (firstList[firstIdx][1] > secondList[secIdx][1]) secIdx += 1
            else firstIdx += 1
        }
        // if (firstIdx == firstList.size)
        return result.toTypedArray()
    }

    fun isOverLap(first: IntArray, second: IntArray): Boolean {
        return !(first[1] < second[0] || first[0] > second[1])
    }

    fun getOverLap(first: IntArray, second: IntArray): IntArray? {
        if (!isOverLap(first, second)) return null
        val start = if (first[0] > second[0]) first[0]
            else second[0]
        val end = if (first[1] > second[1]) second[1]
            else first[1]
        return intArrayOf(start, end)
    }
}
